server/src/main/java/org/elasticsearch/transport/InboundPipeline.java

```diff
@@ -92,99 +90,96 @@ public void handleBytes(TcpChannel channel, ReleasableBytesReference reference)
-                        }
-                    } else {
-                        continueDecoding = false;
-                    }
-                }
-            }
+        if (pending.isEmpty() == false) {
+            // we already have pending bytes, so we queue these bytes after them and then try to decode from the combined message
+            pending.add(reference.retain());
+            doHandleBytesWithPending(channel);
+            return;
+        }
 
-            if (fragments.isEmpty()) {
-                continueHandling = false;
+        while (isClosed == false && reference.length() > 0) {
+            final int bytesDecoded = decode(channel, reference);
+            if (bytesDecoded != 0) {
+                reference = reference.releasableSlice(bytesDecoded);
             } else {
-                try {
-                    forwardFragments(channel, fragments);
-                } finally {
-                    for (Object fragment : fragments) {
-                        if (fragment instanceof ReleasableBytesReference) {
-                            ((ReleasableBytesReference) fragment).close();
-                        }
-                    }
-                    fragments.clear();
-                }
+                break;
             }
         }
+        // if handling the messages didn't cause the channel to get closed and we did not fully consume the buffer retain it
+        if (isClosed == false && reference.length() > 0) {
+            pending.add(reference.retain());
+        }
     }
 
-    private void forwardFragments(TcpChannel channel, ArrayList<Object> fragments) throws IOException {
-        for (Object fragment : fragments) {
-            if (fragment instanceof Header) {
-                assert aggregator.isAggregating() == false;
-                aggregator.headerReceived((Header) fragment);
-            } else if (fragment instanceof Compression.Scheme) {
-                assert aggregator.isAggregating();
-                aggregator.updateCompressionScheme((Compression.Scheme) fragment);
-            } else if (fragment == InboundDecoder.PING) {
-                assert aggregator.isAggregating() == false;
-                messageHandler.accept(channel, PING_MESSAGE);
-            } else if (fragment == InboundDecoder.END_CONTENT) {
-                assert aggregator.isAggregating();
-                try (InboundMessage aggregated = aggregator.finishAggregation()) {
-                    statsTracker.markMessageReceived();
-                    messageHandler.accept(channel, aggregated);
+    private int decode(TcpChannel channel, ReleasableBytesReference reference) throws IOException {
+        return decoder.decode(channel, reference, this::forwardFragment);
+    }
+
+    private void doHandleBytesWithPending(TcpChannel channel) throws IOException {
+        do {
+            final int bytesDecoded;
+            if (pending.size() == 1) {
+                bytesDecoded = decode(channel, pending.peekFirst());
```

Comment:
If the channel is closed then the bytes get released mid-decode. Is that a problem? (possibly not, but I'd like to hear the reasoning)

Reply:
The channel is closed on the same thread that this runs on, we don't release mid-decode. We release after deserializing and handling what we deserialized in all cases.

Reply:
Yeah I really just meant within the `decode()` method.

Reply:
Ah I see. I think there we're good because we don't have any incremental decode on a message. We decode a full message and pass it along to message handling, then return from decode. So closing the channel will always be the last step in `decode` as far as I can see.